### PR TITLE
plan/statistics: Add Statistics types.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @hanfei19910905 @zimulala PTAL

File: plan/statistics/statistics.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (shenli) [inline plan/statistics/statistics.go]: Why put this in plan package? It is nothing to do with plan.

Comment (coocood) [inline plan/statistics/statistics.go]: It will be used in plan to estimate cost.

File: plan/statistics/statistics.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tidb/xapi/tablecodec"
+)
+
+// Column represents statistics for a column.
+type Column struct {
+	ID  int64 // Column ID.
+	NDV int64 // Number of distinct values.
+
+	// Histogram elements.
```
Comment (shenli) [inline plan/statistics/statistics.go]: Please comment more details.

File: plan/statistics/statistics.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	return strings.Join(strs, "\n")
+}
+
+// Table represents statistics for a table.
+type Table struct {
+	info        *model.TableInfo
+	TS          int64 // build timestamp.
+	Columns     []*Column
+	Count       int64
```
Comment (shenli) [inline plan/statistics/statistics.go]: Add some comments for Count and BucketCount.

File: plan/statistics/statistics.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+		} else {
+			if ocurrance == 1 {
+				ocurredOnceCount++
+			}
+			sampleDistinct++
+			ocurrance = 1
+		}
+		lastValue = samples[i]
+	}
+	newValueProbability := float64(ocurredOnceCount) / float64(len(samples))
```
Comment (shenli) [inline plan/statistics/statistics.go]: Do you refer to any paper or source code?

**Comment by coocood** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

File: plan/statistics/statistics.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	Values  []types.Datum
+	Repeats []int64
+}
+
+func (c *Column) String() string {
+	strs := make([]string, 0, len(c.Numbers)+1)
+	strs = append(strs, fmt.Sprintf("column:%d ndv:%d", c.ID, c.NDV))
+	for i := range c.Numbers {
+		strVal, _ := c.Values[i].ToString()
+		strs = append(strs, fmt.Sprintf("num: %d\tvalue: %s\trepeats: %d", c.Numbers[i], strVal, c.Repeats[i]))
```
Comment (zimulala) [inline plan/statistics/statistics.go]: Does it need to add `\t` at the end?

File: plan/statistics/statistics.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	TS          int64 // build timestamp.
+	Columns     []*Column
+	Count       int64 // Total row count in a table.
+	BucketCount int64 // Number of histogram bucket.
+}
+
+// String implements Stringer interface.
+func (t *Table) String() string {
+	strs := make([]string, 0, len(t.Columns)+1)
+	strs = append(strs, fmt.Sprintf("Table:%d ts:%d count:%d", t.info.ID, t.TS, t.Count))
```
Comment (zimulala) [inline plan/statistics/statistics.go]: Does it need to add `\t` or `\n` at the end?

File: plan/statistics/statistics.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+			col.Values = append(col.Values, samples[i])
+			col.Repeats = append(col.Repeats, 0)
+		}
+	}
+	t.Columns[offset] = col
+	return nil
+}
+
+// estimateNDV estimates the number of distinct value given a count and samples.
+// using simplified Good–Turing frequency estimation
```
Comment (zimulala) [inline plan/statistics/statistics.go]: using -> Using?

File: plan/statistics/statistics.go (hunk 5 above)
Comment (coocood) [inline plan/statistics/statistics.go]: No, it will be joined by "\n".

File: plan/statistics/statistics.go (hunk 6 above)
Comment (coocood) [inline plan/statistics/statistics.go]: It will be joined by "\n"

File: plan/statistics/statistics.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+	sampleFactor := t.Count / int64(len(samples))
+	bucketIdx := 0
+	var lastNumber int64
+	for i := int64(0); i < int64(len(samples)); i++ {
+		cmp, err := col.Values[bucketIdx].CompareDatum(samples[i])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if cmp == 0 {
+			// The new item has the same value as current bucket value, to ensures that
```
Comment (zimulala) [inline plan/statistics/statistics.go]: ensures -> ensure ?

File: plan/statistics/statistics.go [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+// estimateNDV estimates the number of distinct value given a count and samples.
+// using simplified Good–Turing frequency estimation
+// See https://en.wikipedia.org/wiki/Good%E2%80%93Turing_frequency_estimation
+func estimateNDV(count int64, samples []types.Datum) (int64, error) {
+	lastValue := samples[0]
+	ocurrance := 1
+	sampleDistinct := 1
+	ocurredOnceCount := 0
```
Comment (zimulala) [inline plan/statistics/statistics.go]: `ocurredOnceCount` -> `occurredOnceCount`

File: plan/statistics/statistics_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			FieldType: *types.NewFieldType(mysql.TypeLonglong),
+		},
+	}
+	tblInfo.Columns = columns
+	timestamp := int64(10)
+	bucketCount := int64(256)
+	t, err := NewTable(tblInfo, timestamp, s.count, bucketCount, [][]types.Datum{s.samples})
+	c.Check(err, IsNil)
+	str := t.String()
+	log.Debug(str)
```
Comment (zimulala) [inline plan/statistics/statistics_test.go]: Remove this line?

Comment (coocood) [inline plan/statistics/statistics_test.go]: It's ok to log it in test.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**